File: go/mfcipher.go
// MFCIPHER - Ternary Huffman Cipher dengan Secret Key
// Alfabet output: 'm' (0), 'f' (1), ' ' (2)
//
// Alur enkripsi:
//   1. Build FREQ dari key (FNV-1a + Fisher-Yates)
//   2. Bangun pohon Huffman terner dari FREQ
//   3. Encode plaintext -> digit terner (via Huffman)
//   4. XOR terner: (digit + keystream) mod 3   <- stream cipher
//
// Penggunaan: ./mfcipher [enc/dec] [input] [output] [key]

package main

import (
	"container/heap"
	"fmt"
	"os"
)

var freqBase = [128]float64{
	1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 5, 1, 1, 2, 1, 1,
	1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
	15, 2, 3, 1, 1, 1, 2, 2, 2, 2, 1, 1, 3, 2, 3, 1,
	4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 1, 1, 1, 1,
	1, 6, 3, 4, 4, 7, 3, 3, 4, 6, 2, 2, 4, 4, 6, 6,
	4, 1, 5, 6, 5, 4, 3, 3, 2, 3, 1, 1, 1, 1, 1, 1,
	1, 9, 2, 5, 6, 12, 3, 3, 5, 9, 1, 1, 6, 4, 8, 8,
	4, 1, 7, 8, 7, 5, 3, 3, 2, 3, 1, 1, 1, 1, 1, 1,
}

// ========== PRNG ==========

func fnv1aHash(key string) uint64 {
	h := uint64(14695981039346656037)
	for i := 0; i < len(key); i++ {
		h ^= uint64(key[i])
		h *= 1099511628211
	}
	return h
}

type xorshift64 struct{ state uint64 }

func newXS(seed uint64) *xorshift64 {
	if seed == 0 {
		seed = 1
	}
	return &xorshift64{state: seed}
}

func (x *xorshift64) next() uint64 {
	x.state ^= x.state << 13
	x.state ^= x.state >> 7
	x.state ^= x.state << 17
	return x.state
}

func buildFreq(key string) [128]float64 {
	freq := freqBase
	rng := newXS(fnv1aHash(key))
	for i := 127; i > 0; i-- {
		j := int(rng.next() % uint64(i+1))
		freq[i], freq[j] = freq[j], freq[i]
	}
	return freq
}

// Keystream menggunakan seed berbeda agar independen dari shuffle
func ksInit(key string) *xorshift64 {
	seed := fnv1aHash(key) ^ 0xdeadbeefcafe
	if seed == 0 {
		seed = 0xdeadbeefcafe
	}
	return &xorshift64{state: seed}
}

// ========== Pohon Huffman ==========

type hnode struct {
	symbol int
	freq   float64
	order  int64
	child  [3]*hnode
}

type nodeHeap []*hnode

func (h nodeHeap) Len() int { return len(h) }
func (h nodeHeap) Less(i, j int) bool {
	if h[i].freq != h[j].freq {
		return h[i].freq < h[j].freq
	}
	return h[i].order < h[j].order
}
func (h nodeHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *nodeHeap) Push(x interface{}) { *h = append(*h, x.(*hnode)) }
func (h *nodeHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func buildTree(freq [128]float64) *hnode {
	h := &nodeHeap{}
	heap.Init(h)
	var counter int64
	for i := 0; i < 128; i++ {
		heap.Push(h, &hnode{symbol: i, freq: freq[i], order: counter})
		counter++
	}
	for (h.Len()-1)%2 != 0 {
		heap.Push(h, &hnode{symbol: -2, freq: 0, order: counter})
		counter++
	}
	for h.Len() > 1 {
		p := &hnode{symbol: -1, order: counter}
		counter++
		for d := 0; d < 3; d++ {
			c := heap.Pop(h).(*hnode)
			p.child[d] = c
			p.freq += c.freq
		}
		heap.Push(h, p)
	}
	return heap.Pop(h).(*hnode)
}

var codes [128][]int8

func extractCodes(n *hnode, path []int8) {
	if n == nil {
		return
	}
	if n.symbol >= 0 && n.symbol < 128 {
		cp := make([]int8, len(path))
		copy(cp, path)
		codes[n.symbol] = cp
		return
	}
	for d := 0; d < 3; d++ {
		extractCodes(n.child[d], append(path, int8(d)))
	}
}

// ========== Encoder & Decoder ==========

func encode(src []byte, key string) []byte {
	ks := ksInit(key)
	out := make([]byte, 0, len(src)*5)
	for _, b := range src {
		sym := int(b)
		if sym >= 128 {
			sym = int('?')
		}
		for _, d := range codes[sym] {
			kd := int(ks.next() % 3)
			cd := (int(d) + kd) % 3
			switch cd {
			case 0:
				out = append(out, 'm')
			case 1:
				out = append(out, 'f')
			default:
				out = append(out, ' ')
			}
		}
	}
	return out
}

func decode(src []byte, root *hnode, key string) []byte {
	ks := ksInit(key)
	out := make([]byte, 0, len(src)/5)
	cur := root
	for i, b := range src {
		var raw int
		switch b {
		case 'm':
			raw = 0
		case 'f':
			raw = 1
		case ' ':
			raw = 2
		default:
			continue
		}
		kd := int(ks.next() % 3)
		d := (raw - kd + 3) % 3
		cur = cur.child[d]
		if cur == nil {
			fmt.Fprintf(os.Stderr, "[-] Data korup pada posisi %d.\n", i)
			return out
		}
		if cur.symbol >= 0 && cur.symbol < 128 {
			out = append(out, byte(cur.symbol))
			cur = root
		}
	}
	return out
}

// ========== Entry Point ==========

func main() {
	if len(os.Args) != 5 {
		fmt.Fprintf(os.Stderr,
			"MFCIPHER - Ternary Huffman Cipher\nPenggunaan: %s [enc/dec] [input] [output] [key]\n",
			os.Args[0])
		os.Exit(1)
	}
	mode, inPath, outPath, key := os.Args[1], os.Args[2], os.Args[3], os.Args[4]
	if mode != "enc" && mode != "dec" {
		fmt.Fprintln(os.Stderr, "[-] Mode tidak valid. Gunakan 'enc' atau 'dec'.")
		os.Exit(1)
	}

	data, err := os.ReadFile(inPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, "[-] Gagal membaca input:", err)
		os.Exit(1)
	}

	freq := buildFreq(key)
	root := buildTree(freq)
	extractCodes(root, nil)

	var result []byte
	if mode == "enc" {
		result = encode(data, key)
		fmt.Fprintf(os.Stderr, "[+] Enkripsi selesai -> %s\n", outPath)
	} else {
		result = decode(data, root, key)
		fmt.Fprintf(os.Stderr, "[+] Dekripsi selesai -> %s\n", outPath)
	}

	if err := os.WriteFile(outPath, result, 0644); err != nil {
		fmt.Fprintln(os.Stderr, "[-] Gagal menulis output:", err)
		os.Exit(1)
	}
}

File: py/mfcipher.py
"""
MFCIPHER - Ternary Huffman Cipher dengan Secret Key
Alfabet output: 'm' (0), 'f' (1), ' ' (2)

Alur enkripsi:
  1. Build FREQ dari key (FNV-1a + Fisher-Yates) untuk 256 simbol byte
  2. Bangun pohon Huffman terner dari FREQ
  3. Prepend header (magic + ekstensi asli) ke plaintext
  4. Encode byte stream -> digit terner (via Huffman)
  5. XOR terner: (digit + keystream) mod 3   <- stream cipher

Penggunaan: python mfcipher.py [enc/dec] [input] [output] [key]
"""

import heapq
import sys
import os
from typing import Optional

# Frekuensi dasar untuk 256 simbol byte
FREQ_BASE: list[float] = [
    # 0x00-0x0F
    1,1,1,1,1,1,1,1,1,2,5,1,1,2,1,1,
    # 0x10-0x1F
    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
    # 0x20-0x2F
    15,2,3,1,1,1,2,2,2,2,1,1,3,2,3,1,
    # 0x30-0x3F
    4,3,3,3,3,3,3,3,3,3,2,2,1,1,1,1,
    # 0x40-0x4F
    1,6,3,4,4,7,3,3,4,6,2,2,4,4,6,6,
    # 0x50-0x5F
    4,1,5,6,5,4,3,3,2,3,1,1,1,1,1,1,
    # 0x60-0x6F
    1,9,2,5,6,12,3,3,5,9,1,1,6,4,8,8,
    # 0x70-0x7F
    4,1,7,8,7,5,3,3,2,3,1,1,1,1,1,1,
    # 0x80-0xFF: byte biner, distribusi flat
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
]

# Header yang disisipkan sebelum enkripsi agar ekstensi asli bisa dipulihkan
HEADER_MAGIC   = b"MFCI"
HEADER_VERSION = 1


# ========== PRNG ==========

def fnv1a_hash(key: str) -> int:
    h = 14695981039346656037
    for ch in key.encode("utf-8"):
        h ^= ch
        h = (h * 1099511628211) & 0xFFFFFFFFFFFFFFFF
    return h


def xorshift64(state: int) -> tuple[int, int]:
    state &= 0xFFFFFFFFFFFFFFFF
    state ^= (state << 13) & 0xFFFFFFFFFFFFFFFF
    state ^= (state >> 7)  & 0xFFFFFFFFFFFFFFFF
    state ^= (state << 17) & 0xFFFFFFFFFFFFFFFF
    return state, state


def build_freq(key: str) -> list[float]:
    freq  = FREQ_BASE[:]
    state = fnv1a_hash(key) or 1
    for i in range(255, 0, -1):
        state, val = xorshift64(state)
        j = val % (i + 1)
        freq[i], freq[j] = freq[j], freq[i]
    return freq


def ks_init(key: str) -> int:
    seed = (fnv1a_hash(key) ^ 0xDEADBEEFCAFE) & 0xFFFFFFFFFFFFFFFF
    return seed or 0xDEADBEEFCAFE


# ========== Pohon Huffman ==========

class Node:
    __slots__ = ("freq", "order", "symbol", "children")

    def __init__(self, freq: float, order: int, symbol: int = -1):
        self.freq     = freq
        self.order    = order
        self.symbol   = symbol
        self.children: list[Optional["Node"]] = [None, None, None]

    def __lt__(self, other: "Node") -> bool:
        return (self.freq, self.order) < (other.freq, other.order)


def build_tree(freq: list[float]) -> Node:
    counter = 0
    pq: list[Node] = []
    for i in range(256):
        heapq.heappush(pq, Node(freq[i], counter, i))
        counter += 1
    while (len(pq) - 1) % 2 != 0:
        heapq.heappush(pq, Node(0.0, counter, -2))
        counter += 1
    while len(pq) > 1:
        children = [heapq.heappop(pq) for _ in range(3)]
        parent = Node(sum(c.freq for c in children), counter)
        parent.children = children
        counter += 1
        heapq.heappush(pq, parent)
    return heapq.heappop(pq)


def extract_codes(node: Node, path: list[int], codes: dict[int, list[int]]) -> None:
    if 0 <= node.symbol < 256:
        codes[node.symbol] = path[:]
        return
    for d, child in enumerate(node.children):
        if child is not None:
            path.append(d)
            extract_codes(child, path, codes)
            path.pop()


# ========== Header ==========

def build_header(ext: str) -> bytes:
    """Buat header yang menyimpan ekstensi asli file.
    Format: MAGIC(4) | VERSION(1) | EXT_LEN(1) | EXT(n)
    """
    ext_bytes = ext.lstrip(".").lower().encode("utf-8")[:31]
    return HEADER_MAGIC + bytes([HEADER_VERSION, len(ext_bytes)]) + ext_bytes


def parse_header(data: bytes) -> tuple[str, bytes]:
    """Ekstrak ekstensi asli dan payload dari hasil dekripsi.
    Return ("", data) jika header tidak ditemukan (file lama tanpa header).
    """
    if not data.startswith(HEADER_MAGIC):
        return "", data
    if len(data) < 6:
        return "", data
    if data[4] != HEADER_VERSION:
        return "", data
    ext_len = data[5]
    if len(data) < 6 + ext_len:
        return "", data
    ext     = data[6:6 + ext_len].decode("utf-8", errors="replace")
    payload = data[6 + ext_len:]
    return ext, payload


# ========== Encoder & Decoder ==========

_D2C = ("m", "f", " ")
_C2D: dict[int, int] = {ord("m"): 0, ord("f"): 1, ord(" "): 2}


def encode(src: bytes, codes: dict[int, list[int]], key: str) -> bytes:
    ks    = ks_init(key)
    parts: list[str] = []
    for b in src:
        for d in codes[b]:
            ks, kd = xorshift64(ks)
            cd = (d + kd % 3) % 3
            parts.append(_D2C[cd])
    return "".join(parts).encode("ascii")


def decode(src: bytes, root: Node, key: str) -> bytes:
    ks  = ks_init(key)
    out = bytearray()
    cur = root
    for i, b in enumerate(src):
        raw = _C2D.get(b)
        if raw is None:
            continue
        ks, kd = xorshift64(ks)
        d = (raw - kd % 3 + 3) % 3
        cur = cur.children[d]
        if cur is None:
            print(f"[-] Data korup pada posisi {i}.", file=sys.stderr)
            return bytes(out)
        if 0 <= cur.symbol < 256:
            out.append(cur.symbol)
            cur = root
    return bytes(out)


# ========== Entry Point ==========

def main() -> None:
    if len(sys.argv) != 5:
        print(
            f"MFCIPHER - Ternary Huffman Cipher\n"
            f"Penggunaan: python {sys.argv[0]} [enc/dec] [input] [output] [key]",
            file=sys.stderr,
        )
        sys.exit(1)

    _, mode, in_path, out_path, key = sys.argv
    if mode not in ("enc", "dec"):
        print("[-] Mode tidak valid. Gunakan 'enc' atau 'dec'.", file=sys.stderr)
        sys.exit(1)

    with open(in_path, "rb") as f:
        data = f.read()

    freq  = build_freq(key)
    root  = build_tree(freq)
    codes: dict[int, list[int]] = {}
    extract_codes(root, [], codes)

    if mode == "enc":
        ext     = os.path.splitext(in_path)[1]
        payload = build_header(ext) + data
        result  = encode(payload, codes, key)
        print(f"[+] Enkripsi selesai -> {out_path}", file=sys.stderr)
    else:
        raw      = decode(data, root, key)
        ext, result = parse_header(raw)
        if ext:
            base, given_ext = os.path.splitext(out_path)
            if not given_ext:
                out_path = base + "." + ext
            print(f"[+] Ekstensi asli dipulihkan: .{ext}", file=sys.stderr)
        print(f"[+] Dekripsi selesai -> {out_path}", file=sys.stderr)

    with open(out_path, "wb") as f:
        f.write(result)


if __name__ == "__main__":
    main()